### ISSUE TITLE
Bug: Excessive CLOSE_WAIT Connections with S3 Plugin

### ISSUE THREAD

**Issue by tvlc** [mariadb-corporation/libmarias3#138]
When using MariaDB with the S3 storage engine implemented via the libmarias3 plugin to interact with a MinIO server (which is behind a HA Proxy), we are experiencing a significant number of TCP connections remaining in the CLOSE_WAIT state on the MariaDB server. During intensive operations, this can result in hundreds or even thousands of CLOSE_WAIT connections, leading to resource exhaustion and Out of Memory (OOM) issues on the MariaDB server. As example, after running a backup of table structures (1000 databases, each with approx. 9 tables with S3 engine) there was 10000, example of netstat command output:

```
Proto Recv-Q Send-Q Local Address           Foreign Address         State       User       Inode      PID/Program name
tcp        0      0 172.20.100.3:39428      95.168.199.230:9443     ESTABLISHED 114        4081627621 3919635/mariadbd    
tcp       25      0 172.20.100.3:35300      95.168.199.230:9443     CLOSE_WAIT  114        4076516828 3919635/mariadbd    
tcp        0      0 172.20.100.3:39362      95.168.199.230:9443     ESTABLISHED 114        4081650053 3919635/mariadbd    
tcp       25      0 172.20.100.3:35192      95.168.199.230:9443     CLOSE_WAIT  114        4076516820 3919635/mariadbd    
tcp        0      0 172.20.100.3:39460      95.168.199.230:9443     ESTABLISHED 114        4081650074 3919635/mariadbd    
tcp        0      0 172.20.100.3:39274      95.168.199.230:9443     ESTABLISHED 114        4081650038 3919635/mariadbd    
tcp       25      0 172.20.100.3:44440      95.168.199.230:9443     CLOSE_WAIT  114        4078342633 3919635/mariadbd    
tcp        0      0 172.20.100.3:39454      95.168.199.230:9443     ESTABLISHED 114        4081650071 3919635/mariadbd    
tcp       25      0 172.20.100.3:50036      95.168.199.230:9443     CLOSE_WAIT  114        4081625332 3919635/mariadbd    
...
``` 

Please notice, that ESTABLISHED connections read data fine, RECV-Q is empty. However, when connection should be closed, 
libmarias3 plugin may not be fully reading incoming data from MinIO before closing the socket. This could lead to sockets lingering in CLOSE_WAIT, consuming system resources, and causing memory issues on the MariaDB server.

Environment:
mariadb-plugin-s3: 10.11.9+maria~ubu2004 amd64
mariadb-server-core: 10.11.9+maria~ubu2004 amd64
mariadb-server: 10.11.9+maria~ubu2004 amd64 
libmariadb3: 10.11.8+maria~ubu2004 amd64

**Comment by LinuxJedi** [comment]
Thanks for the report.

For whoever fixes this (or me if I do it): I suspect the fix would be to add the header `Connection: close` using `CURLOPT_HTTPHEADER`.

**Comment by tvlc** [comment]
Perfect, thank you for prompt response! Do you have any explanation for those unread 25 bytes in RECV-Q?

**Comment by tvlc** [comment]
> Thanks for the report.
> 
> For whoever fixes this (or me if I do it): I suspect the fix would be to add the header `Connection: close` using `CURLOPT_HTTPHEADER`.

Please notice, that the problem look like:
- ha_s3 is not closing socket. A socket in the CLOSE_WAIT state indicates that the remote side has closed the connection (sent a FIN packet), but the local mariadb process has not yet closed the socket. I suspect the server (MariaDB ha_S3 in this case) has received a request to close the connection but hasn’t fully closed it on its end.
- MariaDB is creating new processes and connections even it could reuse existing ones
- There are unread data (exactly 25 bytes) in each RECV-Q of hanging socket

Having this in mind, I am not sure if adding header might have any influence.

**Comment by LinuxJedi** [comment]
Yes, you are correct. I got the wrong end. In theory these should be automatically closed. My guess is it is how the S3 engine is using the library, but some testing will be needed to be sure. If I'm right, either the deinit isn't being called, or something is blocking the connection reuse.

**Comment by tvlc** [comment]
> Yes, you are correct. I got the wrong end. In theory these should be automatically closed. My guess is it is how the S3 engine is using the library, but some testing will be needed to be sure. If I'm right, either the deinit isn't being called, or something is blocking the connection reuse.

Agree. Strange are those unread 25 bytes in RECV-Q. Could these cause such problem?

**Comment by LinuxJedi** [comment]
Without analysing a TCP dump, it would be difficult to see what they are, but my guess is a closing data packet from the other side which hasn't been processed by libcurl in libmarias3 yet.

**Comment by tvlc** [comment]
You are right, TCP dump might be necessary, if the problem would not be obvious to the maintainer. As the connection is https, it would require reconfiguration to htpp to be able to capture the traffic. This is a production system, I will do it within the next maintenance window by end of this month and come back with findings.
